rust/src/pgsql/pgsql.rs

```diff
@@ -244,16 +258,29 @@ impl PgsqlState {
     ///
     /// As Pgsql transactions are bidirectional and may be comprised of several
     /// responses, we must track State progress to decide on tx completion
-    fn is_tx_completed(&self) -> bool {
-        if let PgsqlStateProgress::ReadyForQueryReceived
-        | PgsqlStateProgress::SSLRejectedReceived
-        | PgsqlStateProgress::SimpleAuthenticationReceived
-        | PgsqlStateProgress::SASLAuthenticationReceived
-        | PgsqlStateProgress::SASLAuthenticationContinueReceived
-        | PgsqlStateProgress::SASLAuthenticationFinalReceived
-        | PgsqlStateProgress::ConnectionTerminated
-        | PgsqlStateProgress::Finished = self.state_progress
+    fn is_tx_completed(&self, direction: Direction) -> bool {
```

Comment:
why does updating the per thread/per direction state depend on this global state? Can't we determine this during the parsing of the req/res data itself?

Do understand that a termination has affect on all existing txs

Reply:
For each individual request or response message, the parsing should be enough to decide whether the given request or response message is done. But the completion of a full transaction in Pgsql depends on specific Progress stages - this is what I'm trying to track here. (One good example could be the Simple Query transaction, which could have one request, and then several responses - data description, and then data rows - before we consider the transaction completed.)

Maybe this state/progress is being tracked in the wrong place, and/or maybe (again and still) I misunderstand these in greater depth.

Reply:
I'll keep pondering and analyzing this, to try to reach more clarity.